### PR TITLE
Feature request #7604: use native docker bind mounts with driver-mounts option

### PR DESCRIPTION
**Pull request by g9rga**
Hello everyone! I added new feature for native driver mount **--driver-mounts** by request https://github.com/kubernetes/minikube/issues/7604. 
At this pr, only docker driver supported. Usage example:
`minikube start --driver=docker --driver-mounts=/tmp:/opt/tmp`

Here is test example output:
(⎈ |N/A:default)➜  minikube git:(fix-7604) mkdir /tmp/123
(⎈ |N/A:default)➜  minikube git:(fix-7604) echo "test" > /tmp/123/file.txt
(⎈ |N/A:default)➜  minikube git:(fix-7604) ./out/minikube-linux-amd64 start -p minikube-test --driver=docker --driver-mounts="/tmp/123:/tmp/123:rw"
😄  [minikube-test] minikube v1.12.1 on Arch 20.0.3
    ▪ KUBECONFIG=/home/eth/.kube/config
✨  Using the docker driver based on user configuration
👍  Starting control plane node minikube-test in cluster minikube-test
🚜  Pulling base image ...
💾  Downloading Kubernetes v1.18.3 preload ...
    > preloaded-images-k8s-v4-v1.18.3-docker-overlay2-amd64.tar.lz4: 526.27 MiB
🔥  Creating docker container (CPUs=2, Memory=3900MB) ...
🐳  Preparing Kubernetes v1.18.3 on Docker 19.03.8 ...
🔎  Verifying Kubernetes components...
🌟  Enabled addons: default-storageclass, storage-provisioner
🏄  Done! kubectl is now configured to use "minikube-test"
(⎈ |minikube-test:default)➜  minikube git:(fix-7604) docker inspect -f '{{ .Mounts }}' minikube-test
[{bind  /lib/modules /lib/modules  ro false rprivate} {volume minikube-test /var/lib/docker/volumes/minikube-test/_data /var local z true } {bind  /tmp/123 /tmp/123   true rprivate}]
(⎈ |minikube-test:default)➜  minikube git:(fix-7604) docker exec minikube-test ls /tmp/123
file.txt
(⎈ |minikube-test:default)➜  minikube git:(fix-7604) docker exec minikube-test cat /tmp/123/file.txt                                     
test
(⎈ |minikube-test:default)➜  minikube git:(fix-7604) docker exec minikube-test sh -c "echo 123 > /tmp/123/file.txt"
(⎈ |minikube-test:default)➜  minikube git:(fix-7604) echo /tmp/123/file.txt 
/tmp/123/file.txt
(⎈ |minikube-test:default)➜  minikube git:(fix-7604) 

Screenshot_20200802_111301

### CODE REVIEW

**Comment by afbjorklund** [comment]
It would be nice if we could do something slightly more portable, instead of adding more docker-only flags ? Not _really_ looking forward to duplicating all of these for podman, like we did for docker-env

File: cmd/minikube/cmd/start_flags.go [hunk 1]
```diff
@@ -101,6 +101,7 @@ const (
 	deleteOnFailure         = "delete-on-failure"
 	forceSystemd            = "force-systemd"
 	kicBaseImage            = "base-image"
+	dockerMounts            = "docker-mounts"
```
Comment (medyagh) [inline cmd/minikube/cmd/start_flags.go:104]: any chance we use the same flag as the one we use for the VMs ?

Comment (g9rga) [reply cmd/minikube/cmd/start_flags.go:104]: Do you mean --mount-string option or something else?

**Comment by medyagh** [comment]
I agree with @afbjorklund , we should try to make this not docker secific. on these both PRs

I wonder if we should the existing "-mount-string" option on start? instead of making it docker-specific flag?

File: cmd/minikube/cmd/start_flags.go (hunk 1 above)
Comment (medyagh) [reply cmd/minikube/cmd/start_flags.go:104]: I am open to any solution that doesn't make it docker driver specifi.
--mount-string could be a solution 
or maybe --mount-option

**Comment by g9rga** [comment]
@medyagh @afbjorklund thank you for your proposal. I refactored my PR to use existing option `mount-string`. On minikube start it will use native docker bind mount, instead of 9pfs(mount command). Is it ok?

File: cmd/minikube/cmd/start_flags.go (hunk 1 above)
Comment (Asarew) [reply cmd/minikube/cmd/start_flags.go]: I agree with @afbjorklund statements in the https://github.com/kubernetes/minikube/pull/8159#issuecomment-629603739.
The `--docker-mounts` has different attributes like the options for read-only, rslave, etc. i think adding this logic behind the `--mount-string` won't make the implementation more uniform as the options only applies for the container drivers.

**Comment by ptramsey** [comment]
Hi!  Just wanted to say thank you for doing this work!  Is there anything I can do to help this along?  I'd love to see this landed!

**Comment by g9rga** [comment]
@medyagh @afbjorklund what about **--driver-mounts** option? Format will depend on driver type.

**Comment by medyagh** [comment]
@g9rga do you mind adding an example of using this command in the PR description so I could test it?
or is it possible to add an integration test for it?

**Comment by sharifelgamal** [comment]
@g9rga Could you sign the CLA so we can proceed with testing?

**Comment by g9rga** [comment]
> @g9rga Could you sign the CLA so we can proceed with testing?

I did, `cla/linuxfoundation — g9rga authorized`

**Comment by ptramsey** [comment]
Just checking in here.  :) Looking forward to seeing this!

**Comment by sharifelgamal** [comment]
@g9rga Can you fix the outstanding merge conflict? That way we can rerun the tests and make sure everything is ok.

**Comment by g9rga** [comment]
> @g9rga Can you fix the outstanding merge conflict? That way we can rerun the tests and make sure everything is ok.
@sharifelgamal It fixed. Could you rerun tests?

**Comment by medyagh** [comment]
I think the other Pr might be better, since it uses the already existing flag

https://github.com/kubernetes/minikube/pull/8159/files

File: pkg/drivers/kic/oci/oci.go [hunk 2]
```diff
@@ -592,3 +592,38 @@ func iptablesFileExists(ociBin string, nameOrID string) bool {
 	}
 	return true
 }
+
+func CreateMounts(mountString string) []Mount {
```
Comment (medyagh) [inline pkg/drivers/kic/oci/oci.go:596]: add comment for unexported funcs

File: cmd/minikube/cmd/start_flags.go [hunk 3]
```diff
@@ -196,6 +197,11 @@ func initDriverFlags() {
 	startCmd.Flags().String(hypervVirtualSwitch, "", "The hyperv virtual switch name. Defaults to first found. (hyperv driver only)")
 	startCmd.Flags().Bool(hypervUseExternalSwitch, false, "Whether to use external switch over Default Switch if virtual switch not explicitly specified. (hyperv driver only)")
 	startCmd.Flags().String(hypervExternalAdapter, "", "External Adapter on which external switch will be created if no external switch is found. (hyperv driver only)")
```
Comment (medyagh) [inline cmd/minikube/cmd/start_flags.go:199]: please add a check inside  validateFlags 
to ensure that if user provided this flag with a driver other than Docker it should tell the user this flag wont' work with their driver and they should use --mount-string instead for VM drivers.

**Comment by medyagh** [comment]
@g9rga do you mind trying this PR and see if it works ? and put the output in the PR description ? I did a validation and it didn't seem to be working (if i put a file in the mounted folder I didn't see it in the host)

**Comment by g9rga** [comment]
> @g9rga do you mind trying this PR and see if it works ? and put the output in the PR description ? I did a validation and it didn't seem to be working (if i put a file in the mounted folder I didn't see it in the host)

Did you try it with existing container? Due to docker architecture, mounts only works for new containers(.

File: cmd/minikube/cmd/start_flags.go (hunk 3 above)
Comment (g9rga) [reply cmd/minikube/cmd/start_flags.go:199]: Thank you! I did

File: pkg/drivers/kic/oci/oci.go (hunk 2 above)
Comment (g9rga) [reply pkg/drivers/kic/oci/oci.go:606]: Already added! :+1:

**Comment by mszygenda** [comment]
Would be great if it supported "delegated" / "cached" / "consistent" mount flags which greatly improves performance on MacOS systems.

https://docs.docker.com/docker-for-mac/osxfs-caching/

**Comment by wcjohnson** [comment]
Is there anything that I can do to help with this PR? This functionality is badly needed in the Docker driver. The `minikube mount` 9P file system is far too slow and unstable for anything more than trivial workloads.

**Comment by ptramsey** [comment]
Is there anything that I could do to help this along? I hate to just be sitting on the sidelines cheering!  But I'd really love to see this landed.

**Comment by wcjohnson** [comment]
@g9rga In order to make use of this PR, is it sufficient to just compile and replace the minikube binary?

**Comment by g9rga** [comment]
> @g9rga In order to make use of this PR, is it sufficient to just compile and replace the minikube binary?

Yep, you can compile it with `make darwin/linux/windows` and run compiled binary, for example in my case:
`/home/minikube/out/minikube start --cpus=6 --kubernetes-version=v1.14.10 --driver=docker --memory=6G --driver-mounts="/home/g9rga/www:/home/g9rga/www:rw" --addons=ingress`

**Comment by g9rga** [comment]
I added integration test for new option

**Comment by tstromberg** [comment]
I believe this is obsoleted by #8159 - which does the same thing, but using the existing  --mount-string. My apologies for the confusion: I didn't realize there were two PR's out for the same feature.

**Pull request closed without merging**